src/main/app/controller/courts/AddressController.ts

```diff
@@ -0,0 +1,276 @@
+
+    const pageData: CourtAddressPageData = {
+      addressTypesPrimary: this.getAddressTypesForSelect(addressTypes, true),
+      addressTypesSecondary: this.getAddressTypesForSelect(addressTypes, false),
+      addresses: addresses,
+      writeToUsTypeId: writeToUsTypes.length === 1 ? writeToUsTypes[0] : null,
+      errors: errorMsgs.map(errorMsg => { return { text: errorMsg }; }),
+      fatalError: fatalError,
+      primaryPostcodeInvalid: primaryPostcodeInvalid,
+      secondaryPostcodeInvalid: secondaryPostcodeInvalid,
+      updated: updated
+    };
+
+    res.render('courts/tabs/addressesContent', pageData);
+  }
+
+  private validateAddress(address: DisplayAddress, isPrimaryAddress: boolean): AddressValidationResult {
+    const typeErrors = this.validateType(address, isPrimaryAddress);
+    const addressErrors = this.validateAddressLines(address, isPrimaryAddress);
+    const postcodeErrors = this.validatePostcode(address, isPrimaryAddress);
+    const errorPrefix = isPrimaryAddress ? this.primaryAddressPrefix : this.secondaryAddressPrefix;
+
+    return {
+      postcodeValid: postcodeErrors.length === 0,
+      addressValid: addressErrors.length === 0,
+      errors: typeErrors.concat(addressErrors).concat(postcodeErrors).map(error => errorPrefix + error)
+    };
+  }
+
+  private validateType(address: DisplayAddress, isPrimaryAddress: boolean): string[] {
+    if (isPrimaryAddress || (!!address.postcode || this.addressFieldsNotEmpty(address))) {
+      return !address.type_id ? [this.typeRequiredError] : [];
+    }
+    return [];
+  }
+
+  private validateAddressLines(address: DisplayAddress, isPrimaryAddress: boolean): string[] {
+    const validateSecondaryAddress = (!isPrimaryAddress && (this.addressFieldsNotEmpty(address) || address.postcode?.trim()));
+    const errors: string[] = [];
+
+    if (isPrimaryAddress || validateSecondaryAddress) {
+      if (!address.address_lines?.trim()) {
+        errors.push(this.addressRequiredError);
+      }
+      if(!address.town?.trim()) {
+        errors.push(this.townRequiredError);
+      }
+    }
+    return errors;
+  }
+
+  private validatePostcode(courtAddress: DisplayAddress, isPrimaryAddress: boolean): string[] {
+    const errors: string[] = [];
+    const postcodeRequired = isPrimaryAddress || (!isPrimaryAddress && this.addressFieldsNotEmpty(courtAddress));
+    const postcodeEmpty = !courtAddress?.postcode?.trim();
+
+    if (postcodeRequired && postcodeEmpty) {
+      errors.push(this.postcodeMissingError);
+    }
+
+    if (!postcodeEmpty && !this.postcodeIsValidFormat(courtAddress.postcode)) {
+      errors.push(this.invalidPostcodeError);
+    }
+
+    return errors;
+  }
+
+  private validateAddressTypes(primary: DisplayAddress, secondary: DisplayAddress, writeToUsTypeId: number): string[] {
+    return writeToUsTypeId && !!primary.type_id && !!secondary.type_id && // validate only if types selected in both addresses
```

Comment:
In the `validateType` method, I see the line `return !address.type_id ? [this.typeRequiredError] : [];`, which is run for both the primary and secondary addresses

Is this similar or identical to the logic here: ``!!primary.type_id && !!secondary.type_id` ? 

If so it may be a duplicated check

Reply:
It's similar. It's included here because if either type is unset (in reality, only the secondary could be unset), we can immediately return and avoid any further checking.